network/lite/src/client.rs

```diff
@@ -0,0 +1,210 @@
+use gaffer_udp::GafferPacket;
+use gaffer_udp::non_blocking::GafferSocket;
+use lite_proto::lite::LiteMessage;
+use lite_proto::lite::LiteMessage_Type;
+use protobuf::Message;
+use protobuf;
+use protocol::ClientNegotiationStatus;
+use protocol::ConnectionStatus;
+use protocol::LiteClientEvent;
+use protocol::WriteResult;
+use protocol;
+use std::net::SocketAddr;
+use std::net::ToSocketAddrs;
+use std::time::Duration;
+use std::time::SystemTime;
+
+pub struct LiteClient {
+  internal_socket: GafferSocket,
+  server: SocketAddr,
+  negotiation: Option<ClientNegotiationStatus>,
+  connection: ConnectionStatus,
+}
+
+impl LiteClient {
+  pub fn new<T: ToSocketAddrs>(tsa_self: T, tsa_server: T) -> LiteClient {
+    let gaffer_socket = {
+      let mut socket_addrs = tsa_self.to_socket_addrs()
+        .expect("couldn't coerce to sockets");
+
+      let single_addr = socket_addrs.next()
+        .expect("there was no mapping to addr");
+
+      GafferSocket::bind(single_addr)
+        .expect("couldn't bind to that addr")
+    };
+
+    let server_addr = {
+      let mut socket_addrs = tsa_server.to_socket_addrs()
+        .expect("couldn't coerce to sockets");
+
+      socket_addrs.next().expect("there was no mapping to addr")
+    };
+
+    LiteClient {
+      internal_socket: gaffer_socket,
+      server: server_addr,
+      negotiation: None,
+      connection: ConnectionStatus::Disconnected { id: 0 },
+    }
+  }
+
+  fn syn_bytes() -> Vec<u8> {
+    let mut connect_msg = LiteMessage::new();
+    connect_msg.set_packet_type(LiteMessage_Type::SYN);
+    connect_msg.write_to_bytes().expect("packet wasn't serializable")
+  }
+
+  fn ack_bytes(our_id: u32) -> Vec<u8> {
+    let mut connect_msg = LiteMessage::new();
+    connect_msg.set_packet_type(LiteMessage_Type::ACK);
+    connect_msg.set_origin_id(our_id);
+    connect_msg.write_to_bytes().expect("packet wasn't serializable")
+  }
+
+  fn data_bytes(our_id: u32, payload: [u8; protocol::PAYLOAD_MAX_SIZE]) -> Vec<u8> {
+    let mut connect_msg = LiteMessage::new();
+    connect_msg.set_origin_id(our_id);
+    connect_msg.set_packet_type(LiteMessage_Type::DATA);
+    connect_msg.set_data(payload.to_vec());
+    connect_msg.write_to_bytes().expect("packet wasn't serializable")
+  }
+
+  fn unknown_sender_bytes(&self) -> Vec<u8> {
+    let mut connect_msg = LiteMessage::new();
+    connect_msg.set_packet_type(LiteMessage_Type::UNKNOWN_SENDER);
+    connect_msg.set_origin_id(self.connection.get_id());
+    connect_msg.write_to_bytes().expect("packet wasn't serializable")
+  }
+
+  pub fn connect(&mut self) -> bool {
```

Comment:
I didn't read much, but how often do you expect to not be connected?
Can LiteClient just always be connected? In cases where you don't want to be, maybe Option<LiteClient> could work